### ISSUE TITLE
Serializing a generated function throws odd error

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#33747]
If this is unsupported, it should throw a proper error.

```
julia> @generated function foo()
       end
foo (generic function with 1 method)

julia> serialize(IOBuffer(), first(methods(foo)))
ERROR: type GeneratedFunctionStub has no field inferred
Stacktrace:
 [1] getproperty(::Any, ::Symbol) at ./Base.jl:33
 [2] serialize(::Serializer{Base.GenericIOBuffer{Array{UInt8,1}}}, ::Method) at /home/keno/julia-comparison/usr/share/julia/stdlib/v1.4/Serialization/src/Serialization.jl:415
 [3] serialize(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Method) at /home/keno/julia-comparison/usr/share/julia/stdlib/v1.4/Serialization/src/Serialization.jl:698
 [4] top-level scope at REPL[5]:1
```

**Comment by Keno** [comment]
Note that deserialization is broken also because `MethodInstance` has no `inferred` field:
https://github.com/JuliaLang/julia/blob/master/stdlib/Serialization/src/Serialization.jl#L966-L972

```
julia> fieldnames(Core.MethodInstance)
(:def, :specTypes, :sparam_vals, :uninferred, :backedges, :cache, :inInference)
```

**Comment by jonathanBieler** [comment]
I've hit the same error on v1.5 when trying to serialize a `REPL.REPLCompletions.ModuleCompletion` to a socket, but in v1.7 the error seems a bit different: 

```
using Sockets, Serialization, REPL
listen(2000)
sock = connect(2000)

julia> serialize(sock, REPL.REPLCompletions.ModuleCompletion(Base, "!"))
ERROR: IOError: write: broken pipe (EPIPE)
Stacktrace:
  [1] uv_write(s::TCPSocket, p::Ptr{UInt8}, n::UInt64)
    @ Base ./stream.jl:1010
  [2] unsafe_write(s::TCPSocket, p::Ptr{UInt8}, n::UInt64)
    @ Base ./stream.jl:1064
  [3] unsafe_write
    @ ./io.jl:647 [inlined]
  [4] unsafe_write(s::TCPSocket, p::Base.RefValue{UInt8}, n::Int64)
    @ Base ./io.jl:645
  [5] write
    @ ./io.jl:648 [inlined]
  [6] write(s::TCPSocket, b::UInt8)
    @ Base ./stream.jl:1102
  [7] writetag
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Serialization/src/Serialization.jl:142 [inlined]
  [8] serialize_type(s::Serializer{TCPSocket}, t::DataType, ref::Bool)
    @ Serialization /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Serialization/src/Serialization.jl:574
  [9] serialize_any(s::Serializer{TCPSocket}, x::Any)
    @ Serialization /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Serialization/src/Serialization.jl:645
 [10] serialize
    @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Serialization/src/Serialization.jl:628 [inlined]
 [11] serialize(s::TCPSocket, x::REPL.REPLCompletions.ModuleCompletion)
    @ Serialization /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.7/Serialization/src/Serialization.jl:737
 [12] top-level scope
    @ REPL[5]:1
```

**Comment by Keno** [comment]
That seems like a separate issue. Since you didn't `accept` the connection on the server, it eventually runs out of buffer space. However, it should probably have blocked on the client before that happened.

**Comment by vtjnash** [comment]
https://github.com/JuliaLang/julia/commit/a5758aad30717ae2a71d83718e67579f30e805f6
